### ISSUE TITLE
no detaching the gradients

### ISSUE THREAD

**Issue by xieweiyi** [vfdev-5/BootstrappingLoss#2]
In your bootstrapping loss function, I think you need to stop the gradient propagation for the part where you computing soft or hard pseudo labels from softmax probabilities. otherwise, it does not make sense.

**Comment by vfdev-5** [comment]
@xieweiyi I see your point about interpreting `p` as a pseudo label from the formula:
```
Loss(t, p) = - (beta * t + (1 - beta) * p) * log(p)
```
however, I do not see any hints on that in the paper. The only mention I see there that could be interpreted like that is

> When used with mini-batch stochastic gradient descent, this leads to an EM-like algorithm:  In the E-step, estimate the “true” confidence targets as a convex combination of training labels and model predictions; in the M-step, update the model parameters to better predict those generated targets.

However, on the other hand as authors said that 

> In fact, it can be shown that the resulting objective is equivalent to softmax regression with minimum entropy regularization

And the implementations of entropy regularization I've seen (e.g. [here](https://github.com/ikostrikov/pytorch-a3c/blob/48d95844755e2c3e2c7e48bbd1a7141f7212b63f/train.py#L56)), they do not detach gradients.

**Comment by xieweiyi** [comment]
Hi,

I see your points. But why performing entropy regularization here? as far as I am concerned, entropy regularization is to make predictions more confident. so for a binary problem, p is then either close 1~ or 0~. why it is useful for dealing with noisy labels?

thank you for explaining to me that is a closed form of EM with the use of SGD. never thought about that before. For me, I think fix p makes more sense as we want to use the knowledge learned by the model to provide hints for the true label. I don't think performing entropy regularization makes sense for the noisy label problem.   What do you think?

Best, Xie

**Comment by vfdev-5** [comment]
Hi Xie,

> entropy regularization is to make predictions more confident

Yes, you are right about that. As the paper says :  "Intuitively, minimum entropy regularization encourages the model to have a high confidence in predicting labels"

> why it is useful for dealing with noisy labels?

I think paper's section 3.2 "CONSISTENCY IN MULTI-CLASS PREDICTION VIA BOOTSTRAPPING" could give an answer to that:

> The resulting targets are a convex combination of (1) the noisy training label, and (2) the current prediction of the model.  Intuitively, as the learner improves over time,  its predictions can be trusted more.   This mitigates the damage of incorrect labeling, because incorrect labels are likely to be eventually highly inconsistent with other stimuli predicted to have the same label by the model. By paying less heed to inconsistent labels, the learner can develop a more coherent model, which further improves its ability to evaluate the consistency of noisy labels.

Anyway, I agree that in semi-supervised learning code, we often can see that predictions are considrered as pseudo-labels and are detached to stop gradients propagation. Maybe, it worth to provide an option for that here (in this repo) and run few experiments and see the behaviour...


Victor

**Comment by xieweiyi** [comment]
Hi Victor,

Thanks for the discussion. Indeed, it is worth trying to compare both detach or non-detach way of training to see if this makes a huge difference. 

I will close the issue as your code is correctly implemented as it is according to the original paper.
 
Best, Xie

**Comment by vfdev-5** [comment]
@xieweiyi thanks for raising this question. Actually, digging more inside the code, I think, you are right about stopping gradients as for hard bootstrapping while computing `Loss(t, p) = - (beta * t + (1 - beta) * z) * log(p)` where `z = argmax(p)` as argmax can not be simply differentiated, thus gradient is not propagated. Probably, same logic can be applied to soft bootstrapping.
Anyway, I'll update the code and rerun MNIST experiments with both options.
